File: benchmarks/correctness/python/gc_content.py
"""Compute GC content per sequence from a FASTA file. Output JSON."""
import json
import sys
from Bio import SeqIO

results = {}
for record in SeqIO.parse("data/contigs.fa", "fasta"):
    seq = str(record.seq).upper()
    gc = (seq.count("G") + seq.count("C")) / len(seq) if len(seq) > 0 else 0.0
    results[record.id] = round(gc, 6)

json.dump({"gc_per_sequence": results, "n_sequences": len(results)}, sys.stdout, indent=2)
print()

File: benchmarks/correctness/r/gc_content.R
# Compute GC content per sequence from a FASTA file. Output JSON.
library(Biostrings)
library(jsonlite)

seqs <- readDNAStringSet("data/contigs.fa")
gc_per_seq <- sapply(names(seqs), function(nm) {
    s <- as.character(seqs[[nm]])
    nchar_g <- nchar(gsub("[^G]", "", s))
    nchar_c <- nchar(gsub("[^C]", "", s))
    round((nchar_g + nchar_c) / nchar(s), 6)
})

result <- list(
    gc_per_sequence = as.list(gc_per_seq),
    n_sequences = length(seqs)
)
cat(toJSON(result, auto_unbox = TRUE, pretty = TRUE))
cat("\n")
